### ISSUE TITLE
Reproduce Batch Ensemble

### ISSUE THREAD

**Issue by hwaranlee** [google/edward2#340]
Hi. Thank you for providing this probabilistic package. Recently, I've put on my effort to reproduce the batch ensemble results on cifar-10, and cifar-100 datasets. However, I have difficulty to reproduce as follow:

[CIFAR-100]
| Reported | Batch Size | Learning Rate | Acc1|
Deterministic	?	?	79.8	reported
BatchEnsemble	?	?	81.5	 reported

| Reproduce | Batch Size | Learning Rate | Acc1 |
Deterministic	32	0.025	81.34	
Deterministic	32	0.1	80.64	
Deterministic	128	0.1	80.55
BatchEnsemble	32	0.025	80.76	
BatchEnsemble	32	0.1	79.82

The deterministic model is sometimes even better than the BatchEnsemble model according to the hyper-parameters. I'm working on a single GPU. Can you provide the specific hyper-parameters to reproduce the results?

**Comment by hwaranlee** [comment]
I additionally experimented on CIFAR-100, Wide ResNet 28-10 with BNN-Rank1 algorithm.

| Model | Test Acc (%) | Reported Acc (%) |
| --- |--- |--- |
| Single | 81.02 | 79.8 |
| BatchEnsemble | 81.39 | 81.5 |
| BNN-Rank1 Normal | 80.83 | 81.3 |
| BNN-Rank1 Cauchy | 80.39 | 82.4 |

I did the experiments with the default setting using 2 GPU cores, and I followed the hyperparameters described in the paper:
For normal:
```python experimental/rank1_bnns/cifar.py --ensemble_size 4 --dataset cifar100 --cifar100_c_path '/home/userdata/CIFAR-100-C' --output_dir 'runs/bnnr1-normal' --prior_stddev 0.1 --dropout_rate 1e-3 --random_sign_init -1.0 --corruptions_interval -1 --use_gpu --num_cores 2```
For Cauchy:
```python experimental/rank1_bnns/cifar.py --ensemble_size 4 --dataset cifar100 --cifar100_c_path '/home/userdata/CIFAR-100-C' --output_dir 'runs/bnnr1-cauchy' --alpha_initializer 'trainable_cauchy' --gamma_initializer 'trainable_cauchy' --alpha_regularizer 'cauchy_kl_divergence' --gamma_regularizer 'cauchy_kl_divergence' --prior_stddev 0.01 --dropout_rate 1e-6 --random_sign_init -1.0 --corruptions_interval -1 --use_gpu --num_cores 2```

I tried to find various hyper-parameters that fit to BE and BNN models, but it seems to be very sensitive.. Did I miss something? or the results have high variance? Also I would like to know the std of 10 trials..

**Comment by dustinvtran** [comment]
@ywen666 

Thanks for trying these experiments! The results highly depend on using 8 cores, as that determines the total batch size. If we use fewer cores, then the hyperparameters are no longer tuned, so all bets are off regarding which baseline performs better.

**Comment by ywen666** [comment]
The hparams reported in the paper are specific to 8 cores. Regarding to this set of hparams on BatchEnsemble, the std of 5 trials is within 0.1% on CIFAR-100. BatchEnsemble is sensitive to hyper-parameter selection, especially batch size because it correlates with batch size per ensemble member and number of examples to collect batchnorm statistics. Also on CIFAR-100. we recommend random sign init to be -0.75. We are working on some methods to further improve BatchEnsemble so that it can be robust to hparams choices. 

If I understand correctly, for single gpu, batch size 32 in BatchEnsemble means 8 training examples per ensemble member. Applying a gradient step update based on 8 examples results to a big bias, leading to sub-optimal performance. The additional experiments on two gpus (IIUC, larger batch size per member) improved BatchEnsemble accuracy also supports this. One way to mimic 8 cores is to accumulate gradients, but this requires extra implementation we didn't include in this repo (you also need to correct the number of examples to collect batchnorm statistics as well).

**Comment by dustinvtran** [comment]
hi @hwaranlee! We moved our baselines to https://github.com/google/uncertainty-baselines. If you have Qs, we're happy to answer them further there.
